vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,322 @@
+
+const METADATA_FILE_PARAMETER_PRESENT: u16 = 0x01;
+const METADATA_VIRTUAL_DISK_SIZE_PRESENT: u16 = 0x02;
+const METADATA_VIRTUAL_DISK_ID_PRESENT: u16 = 0x04;
+const METADATA_LOGICAL_SECTOR_SIZE_PRESENT: u16 = 0x08;
+const METADATA_PHYSICAL_SECTOR_SIZE_PRESENT: u16 = 0x10;
+const METADATA_PARENT_LOCATOR_PRESENT: u16 = 0x20;
+
+const METADATA_ALL_PRESENT: u16 = METADATA_FILE_PARAMETER_PRESENT
+    | METADATA_VIRTUAL_DISK_SIZE_PRESENT
+    | METADATA_VIRTUAL_DISK_ID_PRESENT
+    | METADATA_LOGICAL_SECTOR_SIZE_PRESENT
+    | METADATA_PHYSICAL_SECTOR_SIZE_PRESENT;
+
+const METADATA_LENGTH_MAX: u32 = 1024 * 1024; // 1 MiB
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
+    // Block size doesn't meet spec
+    InvalidBlockSize,
+    // Entry size doesn't meet spec
+    InvalidEntryCount,
+    // Logical sector size doesn't meet spec
+    InvalidLogicalSectorSize,
+    // Not a know metadata item
+    InvalidMetadataItem,
+    // Metadata length doesn't meet spec
+    InvalidMetadataLength,
+    // Not a metadata sign
+    InvalidMetadataSign,
+    // Physical sector size doesn't meet spec
+    InvalidPhysicalSectorSize,
+    // Not a valid value
+    InvalidValue,
+    // Not all required metadata are present
+    MissingMetadata,
+    // Failed to read metadata from file
+    ReadMetadata(io::Error),
+    // Found a flag not supported by this implementation
+    UnsupportedFlag,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            InvalidBlockSize => write!(f, "invalid block size count"),
+            InvalidEntryCount => write!(f, "invalid metadata entry count"),
+            InvalidLogicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidMetadataItem => write!(f, "invalid metadata ID"),
+            InvalidMetadataLength => write!(f, "invalid metadata length"),
+            InvalidMetadataSign => write!(f, "metadata sign doesn't match"),
+            InvalidPhysicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidValue => write!(f, "invalid value"),
+            MissingMetadata => write!(f, "not all required metadata found"),
+            ReadMetadata(e) => write!(f, "failed to read metadata headers: {}", e),
+            UnsupportedFlag => write!(f, "this implementation support this metadata flag"),
+        }
+    }
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct DiskSpec {
+    pub disk_id: u128,
+    pub image_size: u64,
+    pub block_size: u32,
+    pub has_parent: bool,
+    pub sectors_per_block: u32,
+    pub virtual_disk_size: u64,
+    pub logical_sector_size: u32,
+    pub physical_sector_size: u32,
+    pub chunk_ratio: u64,
+    pub total_sectors: u64,
+}
+
+struct MetadataTableHeader {
+    _signature: u64,
```

Comment:
```suggestion
    signature: u64,
```

Reply:
We get a warning because we read the signature from the file, make sure if it is the expected one, and store the value in the structure but never use it afterward.

Reply:
With Default(), we do not get the warning!